Sources/SwiftProtobuf/Message+FieldMask.swift

```diff
@@ -0,0 +1,86 @@
+// Sources/SwiftProtobuf/Message+FieldMask.swift - Message field mask extensions
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extend the Message types with FieldMask utilities.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+extension Message {
+
+  /// Checks whether the given path is valid for Message type.
+  ///
+  /// - Parameter path: Path to be checked
+  /// - Returns: Boolean determines path is valid.
+  public static func isPathValid(
+    _ path: String
+  ) -> Bool {
+    Self().hasPath(path: path)
+  }
+
+  internal func isPathValid(
+    _ path: String
+  ) -> Bool {
+    hasPath(path: path)
+  }
+}
+
+extension Message {
+
+  /// Merges fields specified in a FieldMask into another message.
+  ///
+  /// - Parameters:
+  ///   - source: Message should be merged to the original one.
+  ///   - fieldMask: FieldMask specifies which fields should be merged.
+  public mutating func merge(
+    to source: Self,
+    fieldMask: Google_Protobuf_FieldMask
+  ) throws {
+    var copy = self
```

Comment:
Do we need to make the copy, or can we just directly update self?

Reply:
I guess it's needed since we shouldn't update a message unless we make sure our merge won't be thrown. i.e. assume we merge a mask field with 2 paths. The first one performs correctly and when it comes to the second one, it throws an error, so merging happens partially.

Reply:
That does mean this could use a lot of memory depending on the graph size.

The binary format merging doesn't bother to do the buffering here (for memory/performance reasons).  Might be the better choice here as if someone wants to preflight it, they can take the memory hit otherwise, we can error on the side of less memory hit.

@tbkka any thoughts?

Reply:
I changed it and there is no copy anymore. Now we just validate the captured values and throw before any set. Note that our source is `Self` so it should never throw an error (i.e. type mismatch) because merging message fields and source fields are identical.